packages/react/hover-card/src/HoverCard.tsx

```diff
@@ -0,0 +1,289 @@
+import * as React from 'react';
+import { composeEventHandlers } from '@radix-ui/primitive';
+import { useComposedRefs } from '@radix-ui/react-compose-refs';
+import { createContext } from '@radix-ui/react-context';
+import { useControllableState } from '@radix-ui/react-use-controllable-state';
+import * as PopperPrimitive from '@radix-ui/react-popper';
+import { Portal } from '@radix-ui/react-portal';
+import { Presence } from '@radix-ui/react-presence';
+import { Primitive, extendPrimitive } from '@radix-ui/react-primitive';
+import { useId } from '@radix-ui/react-id';
+import { Slot } from '@radix-ui/react-slot';
+import type * as Polymorphic from '@radix-ui/react-polymorphic';
+
+/* -------------------------------------------------------------------------------------------------
+ * HoverCard
+ * -----------------------------------------------------------------------------------------------*/
+
+const HOVERCARD_NAME = 'HoverCard';
+
+type HoverCardContextValue = {
+  triggerRef: React.RefObject<HTMLAnchorElement>;
+  contentId: string;
+  open: boolean;
+  onOpenChange(open: boolean): void;
+  onMouseEnter(): void;
+  onMouseLeave(): void;
+  hasCustomAnchor: boolean;
+  onCustomAnchorAdd(): void;
+  onCustomAnchorRemove(): void;
+};
+
+const [HoverCardProvider, useHoverCardContext] = createContext<HoverCardContextValue>(
+  HOVERCARD_NAME
+);
+
+type HoverCardOwnProps = {
+  open?: boolean;
+  defaultOpen?: boolean;
+  enterDelayDuration?: number;
+  exitDelayDuration?: number;
+  onOpenChange?: (open: boolean) => void;
+};
+
+const HoverCard: React.FC<HoverCardOwnProps> = (props) => {
+  const {
+    children,
+    open: openProp,
+    defaultOpen,
+    onOpenChange,
+    enterDelayDuration = 700,
+    exitDelayDuration = 400,
+  } = props;
+  const triggerRef = React.useRef<HTMLAnchorElement>(null);
+  const [hasCustomAnchor, setHasCustomAnchor] = React.useState(false);
+  const enterEnvoked = React.useRef(false);
+  const enterTimerRef = React.useRef(0);
+  const leaveTimerRef = React.useRef(0);
+
+  const [open = false, setOpen] = useControllableState({
+    prop: openProp,
+    defaultProp: defaultOpen,
+    onChange: onOpenChange,
+  });
+
+  const clearEnterTimer = React.useCallback(() => {
+    clearTimeout(enterTimerRef.current);
+  }, [enterTimerRef]);
+
+  const clearLeaveTimer = React.useCallback(() => {
+    clearTimeout(leaveTimerRef.current);
+  }, [leaveTimerRef]);
+
+  const handleMouseEnter = React.useCallback(() => {
+    clearLeaveTimer();
+    enterTimerRef.current = window.setTimeout(() => {
+      setOpen(true);
+      enterEnvoked.current = true;
+    }, enterDelayDuration);
+  }, [clearLeaveTimer, enterDelayDuration, setOpen]);
+
+  const handelMouseLeave = React.useCallback(() => {
+    clearEnterTimer();
+    leaveTimerRef.current = window.setTimeout(() => {
+      if (enterEnvoked.current) {
```

Comment:
I can see why you asked the question about the state machines.

This kind of logic here definitely makes me think that perhaps we'd have a clearer intent here if we had a state machine instead as this is trying to express the fact that the leave event should be ignored in some situations (ie. if you haven't entered).

With a state machine I believe this would be clearer, because you would only be able leave if you were in open state.

Reply:
Hey @andy-hook, so for this, I think we can probably get away without a state machine as I think we don't need the extra `enterInvokedRef` potentially? Can we rely on the timer? As if there's a timer that kicked off, we know we entered the trigger.

Reply:
There's no way to determine that a timer is running without a flag afaik, we could null the `timerRef` after clearing and check against that but it's arguably more complicated.

Reply:
Took a brief look at using a machine, if we really think it will help the clarity I can make the change but I didn't feel convinced by the extra lines required.